rfc/drafts/mediachain-rfc-3.md

```diff
@@ -0,0 +1,449 @@
+    - Services user search queries.
+    - Asynchronous batching enqueues tasks, until timeout is reached or queue reaches a specified size.
+
++ `Vectorizer`:
+    - High-recall information retrieval model that converts media objects into feature vectors or compact binary codes.
+    - Example training approaches: unsupervised, triplet-based i.e. more-similar pair / less-similar pair.
+    - `Future Indexer` - detect multiple objects per media content and produce multiple feature vectors, possibly annotated with
+       the particular locations at which each object appears in the media.
+    - `Future Indexer` - output multiple vectors corresponding to multiple types of semantic similarity.
+
++ `Vector Compacting`:
+    - `Future Indexer` only: depending on the models chosen, an additional step may be required to effectively transform the feature vectors
+      into more compact binary codes.
+
++ `KNN Index`:
+    - Stores document feature vectors, and does k-nearest-neighbor lookups on those vectors.
+    - `V1 Indexer` will be distributed across worker machines using the "shards & replicas" approach. Queries will be routed to one of the replica sets, and then routed to all shards in that replica set. Results from all shards in the replica set are then combined, ranked, and returned upstream.
+
++ `Dedupe All-vs-All NN`:
+    - Retrieve list of candidate nearest-neighbors media objects, for all media objects.
+
++ `Search Re-Ranking`
+    - `Futuer Indexer` only:
+    - Slower model that ranks all candidate media from the `KNN-Index` lookup, based on relevancy to the query.
+    - Typically involves in-depth attention to each pair of query vs candidate, in order to produce a relevancy score for each candidate media work.
+
++ `Search Override`
+    - `Future Indexer` only:
+    - Able to override potentially poor relevancy output of the underlying search system with known-good answers.
+    - Assists in mitigating certain types of attacks.
+    - Provides possibility for instant updates, versus requiring slow model-retraining before results are visible.
+
++ `Dedupe Pairwise`:
+    - Classification model that inspects all pairs of candidate duplicates returned from the `Dedupe All-vs-All NN` model.
+    - Produces "same" or "different" classifications for all candidate pairs.
+    - Passes classification results to `Dedupe Clustering`.
+
++ `Dedupe Clustering`:
+    - Transforms pairs of same / different classifications from `Dedupe Pairwise` into a maximum-agreement flat clustering.
+
++ `Dedupe Staging`:
+    - Allows bulk approval or denial of newly proposed sets of bulk deduplication updates, prior to being committed to the network.
+    - Filters down the proposed changes to only those that are relevant - those links not already sent to the chain, or the refuting
+      of incorrect links proposed by other entities.
+    - `Future Indexer` will have additional scoring and regression checks here.
+
+
+
+## `Ingestion` Subsystem
```

Comment:
This looks kind of ugy in markdown. Can we drop the code formatting of components in section headings?
Just "Ingestion Subsystem" instead of "`Ingestion` Subsystem"